[JSC] Support JIT compiling for `String#at` by ThunkGenerators

sosukesuzuki opened the pull request:
#### 9d05511754832dd246bb82c366f7619cb88c4f8e
<pre>
[JSC] Support JIT compiling for `String#at` by ThunkGenerators
<a href="https://bugs.webkit.org/show_bug.cgi?id=292474">https://bugs.webkit.org/show_bug.cgi?id=292474</a>

Reviewed by Keith Miller.

This patch changes to support JIT compiling for `String#at`
by ThunkGenerators.

Mircobenchmark results with `useDFGJIT=0`:
                                               TipOfTree                  Patched

string-prototype-at-negative-index           5.6976+-0.1307     ^      4.2518+-0.1516        ^ definitely 1.3400x faster
string-prototype-at-positive-index           6.3179+-1.4795     ^      3.9806+-0.0536        ^ definitely 1.5872x faster

* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::stringCharLoad):
(JSC::charCodeAtThunkGenerator):
(JSC::charAtThunkGenerator):
(JSC::stringAtThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: <a href="https://commits.webkit.org/294841@main">https://commits.webkit.org/294841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e6b9e095c83bab69430a5d1e0e1a66f58dd919b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102961 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108129 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53602 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31136 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78271 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35225 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17782 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92896 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58607 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17617 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10940 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52959 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95636 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87428 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11007 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110502 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101571 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30098 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87258 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30462 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86881 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22164 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31732 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9439 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24366 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30025 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35347 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125204 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29833 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34736 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33160 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31395 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jit/ThunkGenerators.cpp:
```diff
@@ -639,6 +644,23 @@ static void stringCharLoad(SpecializedThunkJIT& jit)
+        // Adjust negative index
+        negativeIndex.link(&jit);
+        // index = length + index
+        jit.add32(SpecializedThunkJIT::regT2, SpecializedThunkJIT::regT1);
```
Why not do 
```
Jump positiveIndex = jit.branch32(MacroAssembler::GreaterThan, SpecializedThunkJIT::regT1, MacroAssembler ::TrustedImm32(0));
// Adjust negative inde: index = length + index
jit.add32(SpecializedThunkJIT::regT2, SpecializedThunkJIT::regT1);
positiveIndex.link(jit);
```

That's one less jump.

sosukesuzuki replied:
Thank you, fixed!

annevk commented on Source/JavaScriptCore/jit/ThunkGenerators.cpp:
```diff
@@ -626,6 +626,11 @@ MacroAssemblerCodeRef<JITThunkPtrTag> stringGetByValGenerator(VM& vm)
+enum class RelativeNegativeIndex : bool {
+    No,
+    Yes,
+};
```
Usually No, Yes enums are a single line.

sosukesuzuki replied:
Thank you, Fixed!

kmiller68 commented on Source/JavaScriptCore/jit/ThunkGenerators.cpp:
```diff
@@ -639,6 +644,13 @@ static void stringCharLoad(SpecializedThunkJIT& jit)
     jit.loadInt32Argument(0, SpecializedThunkJIT::regT1); // regT1 contains the index
 
+    if constexpr (relativeNegativeIndex == RelativeNegativeIndex::Yes) {
+        SpecializedThunkJIT::Jump positiveIndex = jit.branch32(MacroAssembler::GreaterThan, SpecializedThunkJIT::regT1, MacroAssembler ::TrustedImm32(0));
```
I think I said GreaterThan but shouldn’t this be GreaterThanOrEqual? If so, can we add a test?

sosukesuzuki replied:
Thank you, you are right! Fixed!

But we have already tests for `str.at(0)` ( https://github.com/WebKit/WebKit/blob/a36ad67bb5ff6713b4c0dbe49364439a56287f91/JSTests/stress/string-prototype-at-oob.js#L13 ). For such calls, `regT1` becomes equal to `length`, and the subsequent `jit.appendFailure(jit.branch32(MacroAssembler::AboveOrEqual, SpecializedThunkJIT::regT1, SpecializedThunkJIT::regT2));` causes the call to fall back to a non-jit path, so the test could not detect this.

kmiller68 reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Support JIT compiling for `String#at` by ThunkGenerators
https://bugs.webkit.org/show_bug.cgi?id=292474

Reviewed by Keith Miller.

This patch changes to support JIT compiling for `String#at`
by ThunkGenerators.

Mircobenchmark results with `useDFGJIT=0`:
                                               TipOfTree                  Patched

string-prototype-at-negative-index           5.6976+-0.1307     ^      4.2518+-0.1516        ^ definitely 1.3400x faster
string-prototype-at-positive-index           6.3179+-1.4795     ^      3.9806+-0.0536        ^ definitely 1.5872x faster

* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::stringCharLoad):
(JSC::charCodeAtThunkGenerator):
(JSC::charAtThunkGenerator):
(JSC::stringAtThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: https://commits.webkit.org/294841@main
